Copacimportertest

zellerdev opened the pull request:
New PR on a new branch.
Edited Tests. No assignments. Better formatting. Cover DT (documenttype).
New Test in case an empty Text is imported.

Edit: Added the suggestions into the testfile. Checking the equality of Copac and RIS is ongoing.

simonharrer commented:
Can you please have a look at https://ondemand.coverity.com/jobs/uj1ep5hu8118n66eqa61rs5ofc/results/f/4/8/9/620b152144b541f48db807c7eed9.html which indicates issues in the test class. Please close the resources with a try-with-resources block. 

Regarding the `testIsRecognizedFormat`, I think it would be better if there are two lists, one for the accepted and one for the rejected files which can then be iterated and check in a for loop each. This would allow to add all the other files from the other importers that will trickle in over the next month.

zellerdev commented:
Am I assuming right that you mean something like that?
```java
    @Test
    public void testIsRecognizedFormatAccept() throws IOException {
        CopacImporter importer = new CopacImporter();
            LinkedList<String> list = new LinkedList<>();
            list.add("CopacImporterTest1.txt");
            list.add("CopacImporterTest2.txt");

            for (String str : list) {
                try (InputStream is = CopacImporterTest.class.getResourceAsStream(str)) {
                    Assert.assertTrue(importer.isRecognizedFormat(is));
                    is.close();
                }
            }
    }
```

lenhard commented:
Almost :) As you are using try-with-resources, the line `is.close()` is no longer needed. The try block takes care of that automatically.

oscargus commented:
I read somewhere that it is still recommended (by many) to keep the close(), primarily for readability/understanding I guess. Now, I haven't followed it that carefully though...

simonharrer commented:
Do you have a link for that @oscargus ?

oscargus commented:
Maybe not. I tried to find it, but to no avail. I found something which might be what I referred to, but it didn't actually say exactly that. Rather, I found many examples without the close().

simonharrer commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -69,15 +122,55 @@ public void testImportEntries() throws IOException {
+        //CopacImporterTest3.txt is an empty file.
+        List<BibtexEntry> entries = importer.importEntries(
+                CopacImporterTest.class.getResourceAsStream("CopacImporterTest3.txt"), new OutputPrinterToNull());
+        Assert.assertEquals(0, entries.size());
```
When comparing lists, it is better to do an assert that way:

`Assert.assertEquals(Collections.emptyList(), entries);`

Then, the message when the assertion fails is much better, as it contains the entries in the list using their toString method. 

Example:

```java
List<String> values = new LinkedList<>();
values.add("asdf");

Assert.assertEquals(Collections.emptyList(), values);
```

results in 

```
Exception in thread "main" java.lang.AssertionError: expected:<[]> but was:<[asdf]>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at Test.main(Test.java:13)
```

zellerdev replied:
Do you want me to only edit the ```testImportEntries3()``` ? 
Or should I create two Lists on every ```testImport()``` and only make one Assert at the end? For example 
```java
    @Test
    public void testImportEntries() throws IOException {
        Globals.prefs.put("defaultEncoding", "UTF8");

        CopacImporter importer = new CopacImporter();

        List<BibtexEntry> entries = importer.importEntries(
                CopacImporterTest.class.getResourceAsStream("CopacImporterTest1.txt"), new OutputPrinterToNull());
        Assert.assertEquals(1, entries.size());
        BibtexEntry entry = entries.get(0);

        List<String> idealValue = new LinkedList<>();
        idealValue.add("The SIS project : software reuse with a natural language approach");
        idealValue.add("Prechelt, Lutz and Universität Karlsruhe. Fakultät für Informatik");
        idealValue.add("Interner Bericht ; Nr.2/92");
        idealValue.add("1992");
        idealValue.add("Karlsruhe :  Universitat Karlsruhe, Fakultat fur Informatik");
        idealValue.add("Edinburgh");
        idealValue.add("TXT");

        List<String> isValue = new LinkedList<>();
        isValue.add(entry.getField("title"));
        isValue.add(entry.getField("author"));
        isValue.add(entry.getField("series"));
        isValue.add(entry.getField("year"));
        isValue.add(entry.getField("publisher"));
        isValue.add(entry.getField("HL"));
        isValue.add(entry.getField("documenttype"));

        Assert.assertEquals(idealValue, isValue);

        Assert.assertEquals(BibtexEntryTypes.BOOK, entry.getType());
    }
```

simonharrer replied:
I meant only when checking that the size of the list should be 0. You do not get too much information that the list has size 1. It is much easier to debug when you know what values the element in the list has.

zellerdev replied:
Okay. Now I know what you mean. Thank you.
After I have changed ```testImportEntries()``` to the "two list model" I got a failure. I couldn't resolve it because it was one big Message and I didn't saw my mistake. 
This is why I asked before I changed the other one and pushed it.

lenhard commented:
I just realized that a few fixes I implemented when going through coverty warnings this morning (see 5b66b9868e4c7ca97361ba44c2f2f89131828765) conflict with this PR. I fixed the resource leaks and the remainder of this PR is still valuable.

@zellerdev To get it into master, please merge master into your branch. Sorry for the inconvenience, I'll keep my hands off the importers from now.

koppor commented:
I noticed that line 53 reads

    Globals.prefs.put("defaultEncoding", "UTF8");

Could you please change that string to `UTF-8` and check if your test cases still work? - JabRef changed the strings to be more conforming to usual strings - see #155.

Furthermore, could you follow the pattern "If you modify preference, use following pattern:" stated at https://github.com/JabRef/jabref/wiki/Code-Howtos#test-cases?

lenhard commented:
I'd like to add that you should not refer to preferences via their string key, but rather via their static variable in `JabRefPreferences`, e.g.:
```
Globals.prefs.put(JabRefPreferences.DEFAULT_ENCODING, "UTF-8");
```

koppor commented:
I think, it's easier for reviewing if you squash the whole commits into one commit. `git reset master` is your friend.

tobiasdiez commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -6,46 +6,85 @@
+    public ExpectedException thrown = ExpectedException.none();
+
+
     @Test
```
I think it is slightly better to understand if you use `@Test(expect=IOException.class)` instead of the `thrown` construction.

koppor replied:
I agree. In this case, no the whole method may throw an exception and not a single line (as IMHO discussed with @bartsch-dev in another PR)

koppor commented:
Coverage [100%](https://codecov.io/github/JabRef/jabref/src/main/java/net/sf/jabref/importer/fileformat/CopacImporter.java?ref=f3bfd6edcb39b55ce4ef9bd3198112fbe5777abe). :+1:

While wondering why the import source is so small, I checked the header comment http://copac.jisc.ac.uk/faq/#format to check whether there are more fields. There, I found "How do I import Copac records in reference management software? (e.g. Endnote/Zotero)" which contains "If you use Reference Manager you can use the RIS filter to import Copac records."

@obraliar @mairdl @zellerdev @ayanai1 Could you please check whether RIS is really equal to Copac? Can you read each other's code and test cases and report back here?

In case they are equal, I would propose to merge the importer code.

koppor commented:
@zellerdev I can't see what you changed why in your commit. In this case, it would have been better to add another commit. Then, I can easily see, what you changed after my approval. Now, I have to review the whole commit again and I don't know what I have been approved and what not.

zellerdev commented:
I apology for that. Because of the problems building the project I had  in the past this was a routine for me.

tobiasdiez commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -76,17 +97,56 @@ public void testImportEntries() throws IOException {
+            Assert.assertEquals("TXT, PDF", one.getField("documenttype"));
+            Assert.assertEquals(
+                    "Aberdeen ; Birmingham ; Edinburgh ; Trinity College Dublin ; UCL (University College London)",
+                    one.getField("HL"));
```
What is `HL`? From the content, I would suspect it represents the location/city of the publisher. In this case it should  be written to the `address` field in bibtex.

zellerdev replied:
This implementation of CopacImporter knows the tags TI, AU, PY, PU, SE, IS, KW, NT, PD and DT. In this case Copac doesn't know what HL is. So to address this field you have to use the tag HL as field. This ressources already existed. In this case this is a test to watch whether the importer correctly handles with unknown tags.

obraliar commented:
After research we can say that the both file-formats are syntactically not the same. There are a lot of tags (like: SE, PU, PD, NT, HL, ...) in Copac which are note parsable with RISImporter. E.g. `NT` (Copac) <-> `N1` (RIS), which means 'notes'. On closer inspection at the line 163 (else condition) it becomes clear, that Copac accepts all kind of tags which differ from those listed above. That's probably the reason why the code looks short.

koppor commented:
My feeling is that you inspected the code and not the format specifications and concrete results of the exporters linked somewhere from http://copac.jisc.ac.uk/faq/#format. Possibly, it is implicitly clear that the RIS format is completely covered? At #505, I don't see any testfiles of http://copac.jisc.ac.uk.

Is there any Copac format specification? Is there any RIS format specification? My feeling is that tags such as `SE` etc. should be treatable by RIS as well, but I'm not sure.

zellerdev commented:
Updated CopacImporterTest so all checks are succesful and removed conflicts with base branch.

simonharrer commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -75,17 +95,56 @@ public void testImportEntries() throws IOException {
+                    "Aberdeen ; Birmingham ; Edinburgh ; Imperial College ; Liverpool ; Manchester ; Oxford ; Trinity College Dublin",
+                    two.getField("HL"));
+
+            Assert.assertEquals("book", two.getType());
```
Please have a look at how the other successfully merged PRs compared bibentries after the import...

zellerdev commented:
This PR is open since several months now. At this time I used the old method.
I now updated it to use BibtexEntryAssert.

stefan-kolb commented:
@koppor WDYT?

koppor commented on src/test/java/net/sf/jabref/importer/fileformat/SilverPlatterImporterTestNotRecognized.java:
```diff
@@ -25,7 +25,8 @@ public void setUp() throws Exception {
     public final void testIsNotRecognizedFormat() throws Exception {
         List<String> notAccept = Arrays.asList("emptyFile.xml", "IsiImporterTest1.isi",
-                "InspecSilverPlatterImporterTest.txt", "oai2.xml", "RisImporterTest1.ris");
+ "CopacImporterTest1.txt",
```
Please fix the formatting

koppor commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTestFiles.java:
```diff
@@ -0,0 +1,76 @@
+
+            if (size != 1) {
+                for (int i = 1; i <= size; i++) {
+                    BibtexEntryAssert.assertEquals(CopacImporterTest.class, "Bib" + fileName + "-" + i + ".bib",
```
Would it be possible to name the copac files `CopacImporter....txt` and the bib files `CopacImporter....bib`. The prefix `Bib` is not necessary.

Please remove the `Bib` prefix and adapt the filter on line 46: Add `.filter(n -> n.endsWith(".txt)` before `.collect`.

Adapt line 61 to replace `.txt` by `.bib` and just pass `fileName` here.

In case there are multiple entries in your list, possibly BibTexEntryAssert has to be modified. We have to discuss then.

koppor commented on src/test/resources/net/sf/jabref/importer/fileformat/InspecSilverPlatterImporterTest.txt:
```diff
@@ -1,30 +0,0 @@
-Record INSPEC
```
Isn't that file really not used at the InspecImporter or SilverPlatterImportTests?

zellerdev replied:
As far as I know only @tschechlovdev  used it in SilverPlatterImportTest.

koppor replied:
Maybe, it should be renamed then. I wonder, why the test case do not fail when removing that file.

zellerdev replied:
It is used in testIsNotRecognizedFormat. So there is no specific test for the fields. Does this answer your question?

koppor replied:
I thought that Inspec and SilverPlatter are two different formats. Therefore, I would name the file accordingly. - Maybe `SilverPaltterGeneralFile.txt`

koppor replied:
Personal discussion: test case covered by other files. Therefore, OK to remove. @tschechlovdev

simonharrer commented:
Can you please rebase on current master?

koppor commented:
Coverage [98,48%](https://codecov.io/github/JabRef/jabref/src/main/java/net/sf/jabref/importer/fileformat/CopacImporter.java?ref=ede533d62b4ebcf69f2615605622e4dcf0e239c)

[![LGTM](http://lgtm.in/p/KT0PIyUZp)](http://lgtm.in/i/KT0PIyUZp)

simonharrer commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTestFiles.java:
```diff
@@ -0,0 +1,81 @@
+
+            int size = copacEntries.size();
+
+            // workaround because BibtexEntryAssert can only test 1 entry
```
Then provide a method in the BibtexEntryAssert class which can do this, please.

simonharrer commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -1,91 +1,84 @@
-            Assert.assertEquals(2, entries.size());
-            BibEntry one = entries.get(0);
+        List<String> list = testFiles.stream().filter(n -> !n.startsWith("CopacImporterTest"))
+                .filter(n -> !n.startsWith("InspecSilverPlatterImporterTest")).collect(Collectors.toList());
```
Why do you filter out InspecSilverPlatter files as well? This is quite unexpected.

zellerdev replied:
After talking with @tschechlovdev we both agreed to delete the InspecSilverPlatterImporterTest.txt and replace it with another Inspec file which is not recognized by CopacImporter

simonharrer commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -1,91 +1,84 @@
+     */
+    public List<String> getTestFiles() {
+        List<String> files = new ArrayList<>();
+        File d = new File(System.getProperty("user.dir") + "/src/test/resources/net/sf/jabref/importer/fileformat");
```
Please use Java NIO instead. Moreover, you do not need `System.getProperty("user.dir")` as this is the default.

zellerdev replied:
In the latest commit I worked in your feedback except the usage of NIO. I am currently researching it.

zellerdev commented:
Using java.nio for getTestFiles(), removed InspecSilverPlatter and replaced it with another Inspec file which is not recognized by copac and added a new method for BibtexEntryAssert using InputStreams and an ImportFormat as parameter to generate to lists to compare. I hope I solved everything as expected.

simonharrer commented on src/test/java/net/sf/jabref/importer/fileformat/CopacImporterTest.java:
```diff
@@ -1,91 +1,85 @@
+     */
+    public List<String> getTestFiles() throws IOException {
+        List<String> files = new ArrayList<>();
+        Files.newDirectoryStream(Paths.get(FILEFORMAT_PATH)).forEach(n -> files.add(n.getFileName().toString()));
```
a directory stream is a resource which must be closed. use a try-with-resources construct here, please.
